### PR TITLE
8276799: Implementation of JEP 422: Linux/RISC-V Port

### PR DESCRIPTION
**Pull request by yadongw**
This PR backports the implementation of RISC-V port into jdk17u. 

It based on the initial implementation of JEP-422 with some modifications for jdk17u:
1.	The initial patch
    8276799: Implementation of JEP 422: Linux/RISC-V Port
2.	 remove features that do not belong to jdk17u
    revert 8277417: C1 LIR instruction for load-klass
    revert 8278387: Implement UseHeavyMonitors consistently
    revert 8227369: pd_disjoint_words_atomic() needs to be atomic
    revert 8258192: Obsolete the CriticalNatives flag
    revert 8281632: riscv: Improve interpreter stack banging
    revert 8283364: Intrinsify countPositives
3.	add features that belongs to jdk17u
    enable the biased locking
4.	 Wrap modifications to shared code under RISCV and revert changes to other platforms
isolate modification to other platforms
5.	Patches for making compilation pass
8282306: os::is_first_C_frame(frame*) crashes on invalid link access
8282477: vmassert(_last_Java_pc == NULL, "already walkable"); fails with async profiler

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8276799](https://bugs.openjdk.org/browse/JDK-8276799): Implementation of JEP 422: Linux/RISC-V Port
 * [JDK-8282306](https://bugs.openjdk.org/browse/JDK-8282306): os::is_first_C_frame(frame*) crashes on invalid link access
 * [JDK-8282477](https://bugs.openjdk.org/browse/JDK-8282477): [x86, aarch64] vmassert(_last_Java_pc == NULL, "already walkable"); fails with async profiler

### Reviewers
 * [Feilong Jiang](https://openjdk.org/census#fjiang) (@feilongjiang - no project role) ⚠️ Review applies to [47916a22](https://git.openjdk.org/riscv-port-jdk17u/pull/1/files/47916a22eed5c92d56a330a481635c355274062f)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/riscv-port-jdk17u pull/1/head:pull/1` \
`$ git checkout pull/1`

Update a local copy of the PR: \
`$ git checkout pull/1` \
`$ git pull https://git.openjdk.org/riscv-port-jdk17u pull/1/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1`

View PR using the GUI difftool: \
`$ git pr show -t 1`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/riscv-port-jdk17u/pull/1.diff">https://git.openjdk.org/riscv-port-jdk17u/pull/1.diff</a>

</details>

### CODE REVIEW

**Comment by RealFYang** [comment]
@luhenry : That sounds great. It nice if we could have some basic tier1 testing enabled for this branch on GHA.
PS: I believe Yadong is still testing patch and I am also try this on my Unmatched board.

**Comment by yadongw** [comment]
> Love to see this!
> 
> How has it been tested? I'm happy to submit some changes to this riscv-port branch to add testing for tier1 on GHA if that's any help (similar to https://github.com/rivosinc/jdk/actions/runs/4028920698).

Tier1 tests are running on unmatched boards, and a full jtreg tests are running with Qemu. Happy to see more tests, and some failures may be related to existed issues.

**Comment by luhenry** [comment]
Ok, I'll work on submitting a PR tomorrow. It'll be limited to `.github/` so should be fairly easy to isolate.

**Comment by VladimirKempik** [comment]
Hello
could you include https://bugs.openjdk.org/browse/JDK-8293100 into initial backport ?

it was really critical for risc-v and some jni stress tests from vmtestBase/hotspot:tier4 was failing

JDK17 ready backport can be used from https://github.com/syntacore/syntaj17/commit/2053a4ba85078bcfc956676256264f6d7b68061c

**Comment by RealFYang** [comment]
Let me share some of the testing results on my unmatched boards:
- [x] boot-cycle (fastdebug)
- [x] Tier1-3 tests (release)
- [x] Non-trivial benchmark tests like: renaissance, specjbb2015 (release)
- [ ] Tier4 test in progress (release)

**Comment by RealFYang** [comment]
> Hello could you include https://bugs.openjdk.org/browse/JDK-8293100 into initial backport ?
> 
> it was really critical for risc-v and some jni stress tests from vmtestBase/hotspot:tier4 was failing
> 
> JDK17 ready backport can be used from [syntacore/syntaj17@2053a4b](https://github.com/syntacore/syntaj17/commit/2053a4ba85078bcfc956676256264f6d7b68061c)



> Hello could you include https://bugs.openjdk.org/browse/JDK-8293100 into initial backport ?
> 
> it was really critical for risc-v and some jni stress tests from vmtestBase/hotspot:tier4 was failing
> 
> JDK17 ready backport can be used from [syntacore/syntaj17@2053a4b](https://github.com/syntacore/syntaj17/commit/2053a4ba85078bcfc956676256264f6d7b68061c)

I think that deserves another seperate PR after this one? I see the robot is trying to link backporting PRs with the original issue. So it might not be a good idea to put serveral seperate fixes together into one backporting PR.

But the following two changes included in this PR are kind of special as they are needed for a successfull build/bootcycle:
  8282306: os::is_first_C_frame(frame*) crashes on invalid link access
  8282477: vmassert(_last_Java_pc == NULL, "already walkable"); fails with async profiler

**Comment by VladimirKempik** [comment]
> I think that deserves another seperate backporting PR after this one? I see the robot is trying to link each backporting PR with the original issue. So it might not be a good idea to put serveral seperate fixes together into one backporting PR.

Last time I was backporting JEP ( https://github.com/openjdk/jdk11u-dev/pull/715 ) it was ok. Just order scara-bot few commands /issue 8282306 , /issue 8282477.

**Comment by RealFYang** [comment]
> > I think that deserves another seperate backporting PR after this one? I see the robot is trying to link each backporting PR with the original issue. So it might not be a good idea to put serveral seperate fixes together into one backporting PR.
> 
> Last time I was backporting JEP ( [openjdk/jdk11u-dev#715](https://github.com/openjdk/jdk11u-dev/pull/715) ) it was ok. Just order scara-bot few commands /issue 8282306 , /issue 8282477.

Hmm, I think that might be a good way when we are doing the final upstreaming. For now, I still prefer to keep a simple/clean relationship between a backporting PR and the corresponding issue unless there is a special reason.
@yadongw : Please link these two issues (8282306 & 8282477) with this backporting PR.

File: src/hotspot/cpu/riscv/riscv.ad [hunk 1]
```diff
@@ -0,0 +1,10581 @@
...
+    // Handle existing monitor.
+    __ ld(tmp, Address(oop, oopDesc::mark_offset_in_bytes()));
+    __ andi(t0, disp_hdr, markWord::monitor_value);
+    __ bnez(t0, object_has_monitor);
+
+    // Check if it is still a light weight lock, this is true if we
+    // see the stack address of the basicLock in the markWord of the
+    // object.
+
+    __ cmpxchg(/*memory address*/oop, /*expected value*/box, /*new value*/disp_hdr, Assembler::int64, Assembler::relaxed,
```
Comment (feilongjiang) [inline src/hotspot/cpu/riscv/riscv.ad:2422]: Could you adjust indentation here to align with https://github.com/yadongw/riscv-port-jdk17u/blob/47916a22eed5c92d56a330a481635c355274062f/src/hotspot/cpu/riscv/riscv.ad#L2332-L2336

File: src/hotspot/cpu/riscv/riscv.ad [hunk 2]
```diff
@@ -0,0 +1,10581 @@
...
+
+    // Handle existing monitor.
+    __ bind(object_has_monitor);
+    // The object's monitor m is unlocked iff m->owner == NULL,
+    // otherwise m->owner may contain a thread or a stack address.
+    //
+    // Try to CAS m->owner from NULL to current thread.
+    __ add(tmp, disp_hdr, (ObjectMonitor::owner_offset_in_bytes() - markWord::monitor_value));
+    __ cmpxchg(/*memory address*/tmp, /*expected value*/zr, /*new value*/xthread, Assembler::int64, Assembler::aq,
+             Assembler::rl, /*result*/flag); // cas succeeds if flag == zr(expected)
```
Comment (feilongjiang) [inline src/hotspot/cpu/riscv/riscv.ad:2367]: Indentation here could be improved too.

**Review by feilongjiang — APPROVED** [review]
Overall looks good, thanks.

**Comment by RealFYang** [comment]
NOTE: Fix for [JDK-8284068](https://bugs.openjdk.org/browse/JDK-8284068) (https://git.openjdk.java.net/jdk/pull/8055) is also contained in this backport PR.

**Pull request merged**